### PR TITLE
AP_HAL: Improve speed param download in network links (sitl/linux)

### PR DESCRIPTION
**Pull request by Williangalvani**
This improves parameter loading speed with SITL and Linux. After the linked patch, parameter loading speed (not using mavftp) got ~10x slower.

There is no reason for assuming that these are limited to 57600 baud.

In my tests:
param fetch in master with QGC takes ~30 seconds
with this patch (thanks @magicrub), it takes ~2s

Linux shows similar results

~Apparently it was a typo in here:~
https://github.com/ArduPilot/ardupilot/commit/866db281a6de3fce07a5e2214e9893c9aabc8e72#diff-da049e0cf2dd85cd33145ab82688554b96b56444b22e5475a1d961e259aa69d9L136

### CODE REVIEW

**Review by magicrub — APPROVED** [review]
Wowzers! Nice catch. This effects non-ChibiOS and non-MAVFTP param downloads. That's a big deal. This is probably a good one for 4.4.3!

**Comment by WickedShell** [comment]
The change implies a `576000` baud link. Now to be fair SITL should probably sim something faster, but as a generic fall back that's probably faster then we want to keep here. But the linked commit wasn't actually a typo, it was correct.

**Comment by magicrub** [comment]
oh, right! It's bandwidth in bytes so 57.6Kbit BYTE equivalent, minus start bit and a gap, is 57.6k/10 = 5760 BYTES per second.

**Comment by Williangalvani** [comment]
> But the linked commit wasn't actually a typo, it was correct.

Were things incorrect before, then?
I've seen a ~10x slowdown after that patch.
I also assumed changing from `bytes` to `kilobytes` would be a 1000x factor, which is what led me to believe it was a typo.

But yes, I do see that 57600 baud means 5.7kbps

File: libraries/AP_HAL/UARTDriver.h [hunk 1]
```diff
@@ -143,7 +143,7 @@ class AP_HAL::UARTDriver : public AP_HAL::BetterStream {
     virtual uint64_t receive_time_constraint_us(uint16_t nbytes) { return 0; }
 
     virtual uint32_t bw_in_bytes_per_second() const {
-        return 5760;
+        return 57600;
```
Comment (muramura) [inline libraries/AP_HAL/UARTDriver.h:146]: By making it one line, it is the same as any other description.
Only this definition does not appear to need to be on more than one line.

**Comment by WickedShell** [comment]
> So the linked pr should mean we use 30% of the bw at most for params?

I'm not as sure about that. I do know our math for param bandwidth usage is low. Everytime I download params I see a large *reduction* in radio throughput. The same isn't true for the FTP interface, but if you are using the old param interface it's very bad. I'd pretty happily take anything that allocates a larger percentage of bandwidth to params during the download.

**Comment by Williangalvani** [comment]
The current patch worked well here on SITL and Linux.

test in SITL
param fetch with the previous patch (extra 0) took ~3 seconds
param fetch in master takes ~30 seconds
with this patch (thanks @magicrub), it takes ~2s

**Review by magicrub — APPROVED** [review]
LGTM! The code is better in the C file like this but it breaks the pattern from the other AP_HAL implementations. I'm neutral about it. Change it if you want, I'm happy to merge as-is

**Comment by magicrub** [comment]
This has the potential to speed up CI but I don't notice any significant difference.

**Comment by magicrub** [comment]
This PR inspired me to make PR https://github.com/ArduPilot/ardupilot/pull/25365

**Comment by magicrub** [comment]
The reason for this PR makes me think we should change bw_in_bytes_per_second() to be a pure virtual (aka *must* be implemented by the driver)

File: libraries/AP_HAL_Linux/UARTDriver.cpp [hunk 2]
```diff
@@ -443,3 +443,10 @@ uint64_t UARTDriver::receive_time_constraint_us(uint16_t nbytes)
     }
     return last_receive_us;
 }
+
+uint32_t UARTDriver::bw_in_bytes_per_second() const
+{
+    // if connected, assume at least a 10/100Mbps connection
+    const uint32_t bitrate = (_connected && _ip != nullptr) ? 10E6 : _baudrate;
+    return bitrate/10; // convert bits to bytes minus overhead
+};
```
Comment (Ryanf55) [inline libraries/AP_HAL_Linux/UARTDriver.cpp:452]: Add a missing newline, otherwise the pre-commit stuff will start to fail again.

**Review by tridge — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]

**Pull request merged**